m_tunnel_key.go

```diff
@@ -0,0 +1,146 @@
+package tc
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/mdlayher/netlink"
+)
+
+const (
+	tcaTunnelUnspec = iota
+	tcaTunnelKeyTm
+	tcaTunnelKeyParms
+	tcaTunnelKeyEncIPv4Src
+	tcaTunnelKeyEncIPv4Dst
+	tcaTunnelKeyEncIPv6Src
+	tcaTunnelKeyEncIPv6Dst
+	tcaTunnelKeyEncKeyID
+	tcaTunnelKeyPad
+	tcaTunnelKeyEncDstPort
+	tcaTunnelKeyNoCSUM
+	tcaTunnelKeyEncOpts
+	tcaTunnelKeyEncTOS
+	tcaTunnelKeyEncTTL
+)
+
+// TunnelKey contains attribute of the TunnelKey discipline
+type TunnelKey struct {
+	Parms           *TunnelParms
+	Tm              *Tcft
+	KeyEncSrc       *net.IP
+	KeyEncDst       *net.IP	
+	KeyEncKeyID     *uint32
+	KeyEncDstPort   *uint16
+	KeyNoCSUM       *uint8
+	KeyEncTOS       *uint8
+	KeyEncTTL       *uint8
+}
+
+// TunnelParms from from include/uapi/linux/tc_act/tc_tunnel_key.h
+type TunnelParms struct {
+	Index           uint32
+	Capab           uint32
+	Action          uint32
+	RefCnt          uint32
+	BindCnt         uint32
+	TunnelKeyAction uint32
+}
+
+// marshalTunnelKey returns the binary encoding of TunnelKey
+func marshalTunnelKey(info *TunnelKey) ([]byte, error) {
+	options := []tcOption{}
+	if info == nil {
+		return []byte{}, fmt.Errorf("TunnelKey: %w", ErrNoArg)
+	}
+
+	if info.Parms != nil {
+		data, err := marshalStruct(info.Parms)
+		if err != nil {
+			return []byte{}, err
+		}
+		options = append(options, tcOption{Interpretation: vtBytes, Type: tcaTunnelKeyParms, Data: data})
+	}
+	if info.KeyEncSrc != nil {
+		tmp := ipToBytes(*info.KeyEncSrc)
+		options = append(options, tcOption{Interpretation: vtBytes, Type: tcaTunnelKeyEncIPv4Src, Data: tmp})
+	}
+	if info.KeyEncDst != nil {
+		tmp := ipToBytes(*info.KeyEncDst)
+		options = append(options, tcOption{Interpretation: vtBytes, Type: tcaTunnelKeyEncIPv4Dst, Data: tmp})
+	}
+	if info.KeyEncKeyID != nil {
+		options = append(options, tcOption{Interpretation: vtUint32Be, Type: tcaTunnelKeyEncKeyID, Data: *info.KeyEncKeyID})
+	}
+	if info.KeyEncDstPort != nil {
+		options = append(options, tcOption{Interpretation: vtUint16Be, Type: tcaTunnelKeyEncDstPort, Data: *info.KeyEncDstPort})
+	}
+	if info.KeyNoCSUM != nil {
+		options = append(options, tcOption{Interpretation: vtUint8, Type: tcaTunnelKeyNoCSUM, Data: *info.KeyNoCSUM})
+	}
+	if info.KeyEncTOS != nil {
+		options = append(options, tcOption{Interpretation: vtUint8, Type: tcaTunnelKeyEncTOS, Data: *info.KeyEncTOS})
+	}
+	if info.KeyEncTTL != nil {
+		options = append(options, tcOption{Interpretation: vtUint8, Type: tcaTunnelKeyEncTTL, Data: *info.KeyEncTTL})
+	}
+
+	return marshalAttributes(options)
+}
+
+// unmarshalTunnelKey parses the TunnelKey-encoded data and stores the result in the value pointed to by info.
+func unmarshalTunnelKey(data []byte, info *TunnelKey) error {
+	ad, err := netlink.NewAttributeDecoder(data)
+	if err != nil {
+		return err
+	}
+	ad.ByteOrder = nativeEndian
+	for ad.Next() {
+		switch ad.Type() {
```

Comment:
cases `tcaTunnelKeyEncIPv6Src` and `tcaTunnelKeyEncIPv6Dst` are missing.

Reply:
Added the two cases